### PR TITLE
Add callback support to bot::{reply,send}

### PR DESCRIPTION
**Pull request by alFReD-NSH**
* [x] I've read and understood the [Contributing guidelines](./CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](./CODE_OF_CONDUCT.md).
* [x] I've been mindful about doing atomic commits, adding documentation to my changes, not refactoring too much.
* [x] I've a descriptive title and added any useful information for the reviewer. Where appropriate, I've attached a screenshot and/or screencast (gif preferrably).
* [x] I've written tests to cover the new code and functionality included in this PR.
* [x] I've read, agree to, and signed the [Contributor License Agreement (CLA)](https://docs.google.com/a/slack-corp.com/forms/d/1q_w8rlJG_x_xJOoSUMNl7R35rkpA7N6pUkKhfHHMD9c/viewform).

#### PR Summary
This let's `SlackBot::reply` and `SlackBot::send` take an optional callback. While this is not documented, if a callback is passed to Hubot's `Response::send` and `Response::reply`, [Hubot will pass it around.](https://github.com/hubotio/hubot/blob/v3.0.1/src/response.js#L113-L115).

My use case is that, I want to get the `ts` of the message that is sent and then update it later. 

#### Related Issues
N/A

#### Test strategy
Passes the mocha callback to send and reply, so the test will only pass if the callback is called, also checking output to make sure the callback is not related, so not breaking compatibility in any way.

### CODE REVIEW

**Comment by alFReD-NSH** [comment]
This CI is failing since one of the Slack node client dependencies, have published a version that uses ES6 `const` which is I think available only since node 4. I have tested this using node version 8 and everything passes.

File: src/bot.coffee [hunk 1]
```diff
@@ -126,23 +127,32 @@ class SlackBot extends Adapter
...
   send: (envelope, messages...) ->
+    callback = ->
+    if typeof(messages[messages.length - 1]) == 'function'
+      callback = messages.pop()
+
     sent_messages = []
     for message in messages
       if message isnt ''
         sent_messages.push @client.send(envelope, message)
+    Promise.all(messages).then(callback.bind(null, null), callback)
```
Comment (aoberoi) [inline src/bot.coffee]: is this supposed to be `Promise.all(sent_messages)...`?

File: src/bot.coffee [hunk 2]
```diff
@@ -126,23 +127,32 @@ class SlackBot extends Adapter
...
+    callback = ->
+    if typeof(messages[messages.length - 1]) == 'function'
+      callback = messages.pop()
     sent_messages = []
     for message in messages
       if message isnt ''
         message = "<@#{envelope.user.id}>: #{message}" unless envelope.room[0] is 'D'
         @robot.logger.debug "Sending to #{envelope.room}: #{message}"
         sent_messages.push @client.send(envelope, message)
+    Promise.all(messages).then(callback.bind(null, null), callback)
```
Comment (aoberoi) [inline src/bot.coffee]: ditto to above

**Comment by aoberoi** [comment]
@alFReD-NSH thanks for the contribution! i think its a good idea and i'd like to see it merged.

i think something strange is going on with github's "update branch" capability, because i don't see the merge commit being added to this PR after i click that button. the travis build is running against SHA 2428a1190468cdd5ac2ab7c3a342efa0bcdcf20b, which is a 404 in the repo, so the coverage data is not being reported correctly. can you give maintainers permission to update this PR so i can see if i can trigger the right reporting?

**Comment by aoberoi** [comment]
tracking this feature in #473

**Comment by seratch** [comment]
Let me close this PR as #540 has been merged into `master` branch 🎉

**Pull request closed without merging**